internal/auth/authenticator.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2024 EngFlow Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"golang.org/x/oauth2"
+
+	"github.com/EngFlow/auth/internal/autherr"
+	"github.com/EngFlow/auth/internal/browser"
+)
+
+var errUnexpectedHTML = errors.New("request to JSON API returned HTML unexpectedly")
+
+type Backend interface {
+	Authenticate(ctx context.Context, host *url.URL) (*oauth2.Token, error)
+}
+
+type DeviceCode struct {
+	browserOpener browser.Opener
+	clientID      string
+	scopes        []string
+	httpTransport http.RoundTripper
+}
+
+func NewDeviceCode(browserOpener browser.Opener, clientID string, scopes []string) *DeviceCode {
+	return &DeviceCode{
+		browserOpener: browserOpener,
+		clientID:      clientID,
+		scopes:        scopes,
+	}
+}
+
+func (d *DeviceCode) Authenticate(ctx context.Context, host *url.URL) (*oauth2.Token, error) {
+	if d.httpTransport != nil {
+		client := &http.Client{
+			Transport: d.httpTransport,
+		}
+		ctx = context.WithValue(ctx, oauth2.HTTPClient, client)
+	}
+
+	config := &oauth2.Config{
+		ClientID: d.clientID,
+		Scopes:   d.scopes,
+		Endpoint: oauth2.Endpoint{
+			DeviceAuthURL: urlWithPath(host, "api/v1/oauth2/device").String(),
+			TokenURL:      urlWithPath(host, "api/v1/oauth2/token").String(),
+			AuthStyle:     oauth2.AuthStyleInParams,
+		},
+	}
+	res, err := config.DeviceAuth(ctx)
+	if err != nil {
+		if oauthErr := (*oauth2.RetrieveError)(nil); errors.As(err, &oauthErr) {
+			return nil, err
+		}
+		// BUG(CUS-320): Clusters that are not oauth-aware will return HTML with
```

Comment:
btw, is this a bug now?

Reply:
Dunno - do we have any non-oauth-aware backends running, or have we enabled the feature everywhere?

Reply:
It got merged in https://github.com/EngFlow/engflow/pull/22227.

Reply:
The check should stay until *no* backends exist that can return this kind of response

Reply:
We could rephrase to say that older versions of EngFlow redirected unknown API requests to a login page that eventually responded with 200 and HTML. That shouldn't happen anymore, but we can still guard against it. The new behavior is to return 404 instead when OAuth2 device flow is disabled.

Reply:
Ok, rephrased